test/e2e/runner/infra.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+)
+
+// TestnetInfraAPI provides an API for interacting with the infrastructure for
+// an entire testnet.
+type TestnetInfraAPI interface {
+	// GenerateConfig generates any necessary configuration for the
+	// infrastructure provider during testnet setup.
+	GenerateConfig() error
+
+	// Stop attempts to stop the entire testnet.
+	Stop(ctx context.Context) error
+
+	// Pause attempts to pause the entire testnet.
+	Pause(ctx context.Context) error
+
+	// Unpause attempts to resume a paused testnet.
+	Unpause(ctx context.Context) error
+
+	// ShowLogs will print all logs for the whole testnet to stdout.
+	ShowLogs() error
+
+	// TailLogs tails the logs for all nodes in the testnet.
+	TailLogs() error
+
+	// Cleanup stops and destroys all running infrastructure and deletes any
+	// generated files.
+	Cleanup() error
+}
+
+// NodeInfraAPI provides an API for interacting with specific nodes'
+// infrastructure.
+type NodeInfraAPI interface {
+	// ProvisionNode attempts to provision infrastructure for the given node
+	// and starts it.
+	ProvisionNode(ctx context.Context, node *e2e.Node) error
```

Comment:
If we want the division in our infrastructure to have `Provision` and `Start` as different operations, we'll need to adjust the current e2e app docker image that we use.

The `up` command in `docker-compose` will start a running instance of the Tendermint node since the [entrypoint](https://github.com/tendermint/tendermint/blob/deb60d59b5222ecf2369f458f83d08fcbac43ab4/test/e2e/docker/entrypoint#L7) for the image actually runs the process. I'm not sure we really need to maintain two different methods, `Start` and `Provision` when `Start` should really be sufficient. In the DO case, `Start` can bring the host up if it's down and then kickoff the process, whereas in the `docker-compose` case, start will just run the image.

The other option, as far as I can tell, is to have a `Provision` method that is a No-op in the docker case. However, if we're going to do that, I think that this logic should live in the `TestnetInfra` interface as a `Provision` method there since we `Provision` the entire network at once in Terraform, not node-by-node.

Reply:
The only reason I differentiated the two was because the perturbations have a notion of restarting a node that involves stopping and starting it without any change to infra provisioning, but I could just introduce a `RestartNode` function to subsume the two separate operations.

Reply:
Actually, looking at it again now, there are 3 distinct categories of operations in use here:
1. Testnet-wide operations
2. Node-specific operations (including infrastructure)
3. Operations on nodes' processes (not touching the underlying infrastructure)

Reply:
> Actually, looking at it again now, there are 3 distinct categories of operations in use here: ...

I think the distinction makes a bit more sense, although in the case of `docker-compose` the start+provision step are intermixed at the moment. Looks like the API has been updated to reflect that, which seems right.